File: DSA/01_arrays_two_pointers/two_sum.cpp
#include <iostream>
#include <vector>
#include <optional>

/**
 * Finds two numbers in a sorted array that add up to a target.
 * Uses std::optional to safely handle the "not found" state without 
 * relying on magic return numbers (like -1) that can cause silent logic errors.
 */
std::optional<std::pair<int, int>> two_sum_sorted(const std::vector<int>& arr, int target) {
    if (arr.empty()) return std::nullopt;

    size_t left = 0;
    size_t right = arr.size() - 1;

    while (left < right) {
        int current_sum = arr[left] + arr[right];

        if (current_sum == target) {
            return std::make_pair(arr[left], arr[right]);
        } else if (current_sum < target) {
            left++;
        } else {
            right--;
        }
    }
    return std::nullopt;
}

int main() {
    std::vector<int> system_metrics = {10, 20, 35, 50, 75, 80};
    auto result = two_sum_sorted(system_metrics, 70);
    
    if (result) {
        std::cout << "Metrics found: " << result->first << ", " << result->second << "\n";
    }
    return 0;
}

File: DSA/01_arrays_two_pointers/two_sum.rs
/// Finds two numbers in a sorted slice that add up to a target.
/// Rust enforces memory safety without a garbage collector, relying on 
/// ownership rules and built-in Options to handle failure cases securely.
pub fn two_sum_sorted(arr: &[i32], target: i32) -> Option<(i32, i32)> {
    if arr.is_empty() {
        return None;
    }

    let mut left = 0;
    let mut right = arr.len() - 1;

    while left < right {
        let current_sum = arr[left] + arr[right];

        if current_sum == target {
            return Some((arr[left], arr[right]));
        } else if current_sum < target {
            left += 1;
        } else {
            right -= 1;
        }
    }
    
    None
}

fn main() {
    let telemetry_data = vec![10, 20, 35, 50, 75, 80];
    
    // Pattern matching forces the engineer to handle the failure state
    match two_sum_sorted(&telemetry_data, 70) {
        Some((val1, val2)) => println!("Telemetry targets found: {}, {}", val1, val2),
        None => println!("No matching targets found."),
    }
}
